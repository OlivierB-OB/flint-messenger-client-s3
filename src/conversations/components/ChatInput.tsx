import Grid from '@material-ui/core/Grid';
import IconButton from '@material-ui/core/IconButton';
import TextField from '@material-ui/core/TextField';
import Send from '@material-ui/icons/Send';
import React, { useEffect } from 'react';
import { connect } from 'react-redux';
import { Action } from 'redux';
import { ThunkDispatch } from 'redux-thunk';
import { IAppState } from '../../appReducer';
import { makeSendMessage } from '../actions/makeSendMessage';
import { updateMessageEdition } from '../actions/updateMessageEdition';

interface IChatInputDisplayProps {
  conversationId: string;
  messageEdition: string;
  updateMessageEdition: (text: string) => void;
  sendMessage: (conversationId: string) => void;
}

function ChatInputDisplay({
  messageEdition,
  conversationId,
  updateMessageEdition,
  sendMessage,
}: IChatInputDisplayProps) {
  useEffect(() => updateMessageEdition(''), [updateMessageEdition, conversationId]);
  return (
    <Grid container justify="center">
      <Grid item xs={11} justify="center">
        <TextField
          fullWidth={true}
          value={messageEdition}
          variant="filled"
          onChange={(event) => updateMessageEdition(event.target.value)}
          onKeyDown={(event) => event.key === 'Enter' && sendMessage(conversationId)}
        />
      </Grid>
      <Grid item container xs={1} justify="center" alignItems="center">
        <IconButton aria-label="contacts" onClick={() => sendMessage(conversationId)}>
          <Send fontSize="large" color="primary" />
        </IconButton>
      </Grid>
    </Grid>
  );
}

export interface IChatInputProps {
  conversationId: string;
}

const mapStateToProps = ({ conversations }: IAppState, { conversationId }: IChatInputProps) => ({
  conversationId,
  messageEdition: conversations.messageEdition,
});

const mapDispatchToProps = (dispatch: ThunkDispatch<IAppState, void, Action>) => ({
  updateMessageEdition: (text: string) => void dispatch(updateMessageEdition(text)),
  sendMessage: (conversationId: string) => void dispatch(makeSendMessage(conversationId)),
});

export const ChatInput = connect(mapStateToProps, mapDispatchToProps)(ChatInputDisplay);
