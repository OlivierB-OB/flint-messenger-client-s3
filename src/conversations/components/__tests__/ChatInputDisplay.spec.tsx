import IconButton from '@material-ui/core/IconButton';
import TextField from '@material-ui/core/TextField';
import React from 'react';
import { shallow, mount } from 'enzyme';
import { ChatInputDisplay } from '../ChatInput';

describe('ChatInputDisplay', () => {
  it('should display the chat input form', async () => {
    const props: any = {
      conversationId: '123',
      messageEdition: 'message edition',
      updateMessageEdition: jest.fn(),
      sendMessage: jest.fn(),
    };
    const component = shallow(<ChatInputDisplay {...props} />);
    expect(component).toMatchSnapshot();
  });

  it('should allow editing the message', async () => {
    const props: any = {
      conversationId: '123',
      messageEdition: 'message edition',
      updateMessageEdition: jest.fn(),
      sendMessage: jest.fn(),
    };
    const component = shallow(<ChatInputDisplay {...props} />);
    component.find(TextField).simulate('change', { target: { value: 'new message value' } });
    expect(props.updateMessageEdition).toHaveBeenCalledWith('new message value');
  });

  it('should allow sending the message', async () => {
    const props: any = {
      conversationId: '123',
      messageEdition: 'message edition',
      updateMessageEdition: jest.fn(),
      sendMessage: jest.fn(),
    };
    const component = shallow(<ChatInputDisplay {...props} />);
    expect(component.find(IconButton).at(0).prop('type')).toEqual('submit');
    component.find('form').simulate('submit', { preventDefault: () => null });
    expect(props.sendMessage).toHaveBeenCalledWith('123');
  });

  it('should reset the message edition when loaded', async () => {
    const props: any = {
      conversationId: '123',
      messageEdition: 'message edition',
      updateMessageEdition: jest.fn(),
      sendMessage: jest.fn(),
    };
    mount(<ChatInputDisplay {...props} />);
    expect(props.updateMessageEdition).toHaveBeenCalledWith('');
  });
});
