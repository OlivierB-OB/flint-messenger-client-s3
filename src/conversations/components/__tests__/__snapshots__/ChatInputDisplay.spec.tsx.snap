// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ChatInputDisplay should display the chat input form 1`] = `
<form
  onSubmit={[Function]}
>
  <WithStyles(ForwardRef(Grid))
    alignItems="center"
    container={true}
    justify="center"
  >
    <WithStyles(ForwardRef(Grid))
      item={true}
      xs={11}
    >
      <WithStyles(ForwardRef(TextField))
        fullWidth={true}
        onChange={[Function]}
        value="message edition"
        variant="filled"
      />
    </WithStyles(ForwardRef(Grid))>
    <WithStyles(ForwardRef(Grid))
      item={true}
      xs={1}
    >
      <WithStyles(ForwardRef(IconButton))
        aria-label="contacts"
        type="submit"
      >
        <SendIcon
          color="primary"
          fontSize="large"
        />
      </WithStyles(ForwardRef(IconButton))>
    </WithStyles(ForwardRef(Grid))>
  </WithStyles(ForwardRef(Grid))>
</form>
`;
